src/Servers/Kestrel/Core/src/Internal/Infrastructure/StringUtilities.cs

```diff
@@ -118,39 +118,35 @@ public static unsafe bool TryGetAsciiString(byte* input, char* output, int count
         /// <param name="separator"></param>
         /// <param name="number"></param>
         /// <returns></returns>
-        public static unsafe string ConcatAsHexSuffix(string str, char separator, uint number)
+        public static string ConcatAsHexSuffix(string str, char separator, uint number)
         {
             var length = 1 + 8;
             if (str != null)
             {
                 length += str.Length;
             }
 
-            // stackalloc to allocate array on stack rather than heap
-            char* charBuffer = stackalloc char[length];
-
-            var i = 0;
-            if (str != null)
+            return string.Create(length, (str, separator, number), (charBuffer, tuple) =>
             {
-                for (i = 0; i < str.Length; i++)
+                var i = 0;
+                if (tuple.str != null)
                 {
-                    charBuffer[i] = str[i];
+                    for (i = 0; i < tuple.str.Length; i++)
+                    {
+                        charBuffer[i] = tuple.str[i];
```

Comment:
Maybe `tuple.str.AsSpan().CopyTo(charBuffer)` instead of the manual copy-loop?
`i` needs to be incremented by `tuple.str.Length` then.

Here `charBuffer` is used, other places use `span`. I would unify this to `charBuffer` or `buffer`.